README.md

```diff
@@ -216,34 +215,109 @@ If you register the service providers in the correct order in your container (A
 - [Stratigility Module](https://github.com/thecodingmachine/stratigility-harmony): A service provider for the Stratigility PSR-7 middleware.
 - [Whoops PSR-7 Middleware Module](https://github.com/thecodingmachine/whoops-middleware-universal-module): a service provider for the [Whoops](https://filp.github.io/whoops/) [PSR-7 middleware](https://github.com/franzliedke/whoops-middleware).
 
-## FAQ
+## Best practices
+
+### Playing nice with all containers
+
+The goal of this project is to offer a viable solution for **agnostic container configuration**. There is a wide variety of containers, and they can have very different behaviour.
+In regard to service providers, it is important to understand there are 2 classes of containers:
+
+- *runtime containers*
+- *compiled/cached containers*
+
+Runtime containers will typically instantiate service providers on each request. The `getServices` method will be called on each request to configure the container.
+
+By contrast, cached/compiled container will avoid calling `getServices` on each call by caching the results or dumping a PHP file containing the container. Therefore, service provider writers should not rely on the `getServices` method being always called before the factories.
+
+A good approximation is to say that service providers SHOULD be immutable. Once initialized, the state of the service provider should not change. 
```

Comment:
I think all that section can be removed, the README is already super long I'd rather keep it as simple as we can. In the end the information here is:

> service provider writers should not rely on the `getServices` method being always called before the factories.

And I don't think it's crucial, if it affected anyone it means there are doing something really wrong.